[CK_BUILDER] Factory tests

Snektron opened the pull request:
## Proposed changes

This pull requests adds some initial "factory tests" - these check that the instances which are used in MIOpen are actually present in CK. The main reason for this is documentation and sanity checking. Its likely that these tests get outdated fast, so we'll have to maintain them, but fortunatenly this is quite straight forward and shouldn't take a lot of time once they are in place.

## Design

The main idea is straight forward: Each test corresponds to one or a few "device operations" (structs which specialize `DeviceOperationInstanceFactory`), and checks whether all instances defined in that factory are present. To this end, I've added some utilities so that the test file itself is very easy to write:
- `InstanceSet` represents a set of instances, either obtained from a factory, or from a statically defined list. `InstanceSet::from_factory<DeviceOp>` can be used to query `DeviceOperationInstanceFactory` directly, while the initializer-list constructor can be used for statically defined list of (expected) instances. Since we don't really care about order (in the future we might have an implementation in CK builder that returns the instances in a different order), these instances are stored internally in an `std::set<std::string>`. This makes it easy to compare two sets for items which are present or not.
- `InstanceMatcher` is a GoogleTest utility that can be used to compare two `InstanceSet`s. If you write `EXPECT_THAT(actual, InstancesMatch(expected))`, you will get a readable error message explaining which instances were either missing or not present.

So, the basic implementation of a test is then consists of the following parts. For each `DeviceOperationInstanceFactory<T>` specialization which is relevant to MIOpen:

- Create a file thats named `test_ck_factory_something_thats_named_after_the_specific_specialization_you_are_testing.cpp`
- In it, include the header that the specialization is defined it.
- I like to define some utilities, imports, etc. CK namespaces are long, so i just use `using ck::Thing` to make the code more readable
- Define a test fixture as follows:
```c++
template <typename Case>
struct CkFactoryTestConvFwd : public testing::Test
{
    static auto get_actual_instances()
    {
        return InstanceSet::from_factory<typename Case::DeviceOp>();
    }

    static auto get_expected_instances() { return InstanceSet(Case::expected); }
};
```
- Define each of the test cases as follows:
```c++
struct SomeRelatableName
{
    using DeviceOp = DeviceGroupedConvFwdMultipleABD<...>;
    
    constexpr static auto expected = {
        // clang-format off
        ""
        // clang-format on
    };
}
```
- Finally, define the actual test.
```
using TestTypes = ::testing::Types<SomeRelatableName, ... etc>;

TYPED_TEST_SUITE(MyTestSuite, TestTypes);

TYPED_TEST(MyTestSuite, TestInstances)
{
    auto actual   = TestFixture::get_actual_instances();
    auto expected = TestFixture::get_expected_instances();

    EXPECT_THAT(actual, InstancesMatch(expected));
}
```
- Compile and run the test, then update the `expected` items with the item that the error tells you to.

Additional notes:
- When a particular kernel does not have reflection implemented yet, it will return "". So, any time that you see "" in the expected list of instances, a reflection implementation is missing. During an error this will show up as `(empty string)`.
- A test suite should cover all the combinations of types/parameters present in the relevant specialization of `DeviceOperationInstanceFactory`.
- If we expect no instance, `auto expected = {};` will not compile. In those cases I usually put `std::initializer_list<const char*> expected = {};`.

## Implementation

This pull request implements the scaffolding (including tests for the utilities outlined above) and all implementations for `DeviceGroupedConvFwdMultipleABD`, except the quantized ones (these do not seem to be used by MIOpen). You can verify this for some part by running the following (rg is [ripgrep](https://github.com/BurntSushi/ripgrep)):
```
$ rg -l -w DeviceGroupedConvFwdMultipleABD library/include/ck/library/tensor_operation_instance/gpu | rg .hpp | rg -v /quantization/ | sort -u
```

<details>

```
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bias_bnorm_clamp.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bias_clamp.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bilinear.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_clamp.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convinvscale.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale_add.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale_relu.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_dynamic_op.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scaleadd_ab.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scaleadd_scaleadd_relu.hpp
library/include/ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scale.hpp
```

</details>

and comparing this with the implementation:
```
$ rg -woNI 'ck/library/tensor_operation_instance/gpu/.*\.hpp' experimental/ | sort -u
```

<details>

```
ck/library/tensor_operation_instance/gpu/grouped_conv_fwd/device_grouped_conv_fwd_xdl_comp_instance.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bias_bnorm_clamp.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bias_clamp.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_bilinear.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_clamp.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convinvscale.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale_add.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_convscale_relu.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_dynamic_op.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scaleadd_ab.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scaleadd_scaleadd_relu.hpp
ck/library/tensor_operation_instance/gpu/grouped_convolution_forward_scale.hpp
```

</details>

The other part is manually verifying that every combination of parameters that exist in the tensor operation instances are covered by the tests. Unfortunately there is not an easy way to check whether something is missing. Neither is there a way to detect when a new valid combination of parameters is added (for example, when a new layout is supported).

#### Fixes

We already found a few issues which we've fixed. I realize that this means that those headers were probably never used and so the relevant test is probably quite useless, but at least its sort of serving its purpose already.
- Fixed a comment about the layout in `grouped_convolution_forward.hpp`
- Fixed an incorrect header + error in template declaration in `grouped_convolution_forward_dynamic_op.hpp`
- Fixed an incorrect prototype in `grouped_convolution_forward_scale.hpp`
- Fixed the interpreter path in `script/remove_exec_bit.sh`. `/bin/bash` is not POSIX standard (I'm working on NixOS so this errored for me...)

Snektron committed:
ck-builder: add InstanceSet and InstanceMatcher

These types can be used to check whether a list of instances
defined statically/obtained from a DeviceOp is congruent. We don't
care about the order, just which items are present or not. Note that
when GetInstanceString() is not implemented for a particular kernel,
it returns "" (via the base class implementation). When this is the
case, tests should temporarily add "" to the expected list of
instances until the implementation is done.

Snektron commented:
To be rebased when https://github.com/ROCm/composable_kernel/pull/3075 is merged

kabrahamAMD commented on experimental/builder/test/testing_utils.cpp:
```diff
@@ -0,0 +1,86 @@
+#include "testing_utils.hpp"
```
License

Snektron replied:
Whoops, fixed. Also rebased on #3075

aosewski commented on experimental/builder/test/CMakeLists.txt:
```diff
@@ -19,5 +19,15 @@ add_ck_builder_test(test_conv_builder
     test_instance_traits.cpp)
 
+add_ck_builder_test(test_testing_utils test_testing_utils.cpp)
+target_link_libraries(test_testing_utils PRIVATE composablekernels::device_conv_operations)
```
You can reuse for this your function `add_ck_factory_test` defined below.

Snektron replied:
Done

Snektron committed:
ck-builder: ck factory convscale relu/add

bartekxk commented on experimental/builder/include/ck_tile/builder/reflect/instance_traits_device_conv2d_fwd_xdl_c_shuffle_nhwc_kyxc_nhwk.hpp:
```diff
@@ -0,0 +1,285 @@
+#pragma once
+
+#include "instance_traits.hpp"
+
```
Should we add support for non grouped conv in the builder? These instances are getting deprecated

bartekxk replied:
I recommend to drop these instances

Snektron replied:
Done. I've replaced them with some other instances that I was working on, though not all of these have reflection yet.

bartekxk commented on experimental/builder/include/ck_tile/builder/reflect/instance_traits_device_conv2d_fwd_xdl_nhwc_kyxc_nhwk.hpp:
```diff
@@ -0,0 +1,258 @@
+          ck::index_t BBlockTransferSrcVectorDim,
+          ck::index_t BBlockTransferSrcScalarPerVector,
+          ck::index_t BBlockTransferDstScalarPerVector_K1,
+          bool BBlockLdsAddExtraN,
```
The same

Snektron committed:
ck-builder: ck factory grouped conv fwd

Snektron committed:
ck-builder: ck factory grouped conv fwd scaleadd ab

Snektron committed:
ck-builder: ck factory grouped conv fwd scale

kabraham-streamhpc committed:
ck-builder: ck factory grouped conv fwd convscale

implemented tests for instances from grouped_convolution_forward_convscale.hpp:100

kabraham-streamhpc committed:
ck-builder: ck factory grouped conv fwd combconvscale

implemented tests for instances from grouped_convolution_forward_convscale.hpp:210

Snektron committed:
ck-builder: ck factory grouped conv fwd convinvscale

kabraham-streamhpc committed:
ck-builder: added grouped_conv_bilinear to tests

Snektron committed:
fix interpreter path on remove_exec_bit script

/bin/bash does not need to exist on POSIX systems.

Snektron committed:
ck-builder: ck factory grouped conv fwd bias clamp

Snektron committed:
ck-builder: ck factory grouped conv fwd scaleadd scaleadd relu

Snektron committed:
ck-builder: ck factory grouped conv fwd bias bnorm clamp

kabraham-streamhpc committed:
ck-builder: ck factory grouped conv fwd dynamic op + fixes

kabrahamAMD committed:
ck-builder: added test_grouped_convolution_forward_clamp

shumway commented on experimental/builder/test/test_ck_factory_grouped_convolution_forward.cpp:
```diff
@@ -0,0 +1,2345 @@
+    using DeviceOp = ::DeviceOp<2, NHWGC, GKYXC, NHWGK, float>;
+
+    constexpr static auto expected = {
+        // clang-format off
```
nit: Isn't it better to just use clang format and have it wrap the long string over a couple line. I'd prefer to avoid `clang-format off`.

Snektron replied:
I don't agree. Without wrapping its much easier to see which lines form a particular instance string in my opinion, and for example to use multi-line editing to change something. It also makes it easier to write tooling that edits the set of expected instances (like my python script which automatically updates the results).

shumway replied:
Ah, I see now. Yes, your choice is definitely much better.

kabraham-streamhpc committed:
ck-builder: fixed tests for grouped_convolution_forward_clamp.cpp
